rfcs/0000-safe-transmute.md

```diff
@@ -0,0 +1,2183 @@
+where
+    T: TransmuteInto<NonEmptySlice<'static, u8>, NeglectStability>,
+{
+    // unsafe because we `NeglectStability`
+    unsafe { src.unsafe_transmute_into() }
+}
+```
+
+Demoting `NeglectStability` to unsafe-status does not stop type authors from opting-in to stable (and thus safe) transmutations; e.g., with `derive(PromiseTransmutableFrom)`:
+```rust
+pub mod crate_a {
+
+    #[derive(PromiseTransmutableFrom)]
+    #[repr(C)]
+    pub struct NonEmptySlice<'a, T>(pub private::NonEmptySliceInner<'a, T>);
+
+    impl<'a, T> NonEmptySlice<'a, T> {
+        pub fn from_array<const N: usize>(arr: &'a [T; N], len: usize) -> Self {
+            assert!(len <= N && len > 0);
+            Self(
+                private::NonEmptySliceInner {
+                    data: arr as *const T,
+                    len,
+                    lifetime: core::marker::PhantomData,
+                }
+            )
+        }
+
+        pub fn first(&self) -> &'a T {
+            unsafe { &*self.0.data }
+        }
+    }
+
+    // introduce a private module to avoid `private_in_public` error (E0446):
+    pub(crate) mod private {
+        #[derive(PromiseTransmutableFrom)]
+        #[repr(C)]
+        pub struct NonEmptySliceInner<'a, T> {
+            pub data: *const T,
+            pub len: usize,
+            pub lifetime: core::marker::PhantomData<&'a ()>,
+        }
+    }
+
+}
+```
+In the above example, the type author declares `NonEmptySlice` and `NonEmptySliceInner` to be stably instantiatable via transmutation. Given this, a third-party no longer needs to resort to `unsafe` code to violate the the invariants on `inner`:
+```rust
+pub evil_constructor<T>(src: T) -> NonEmptySlice<'static, u8>
+where
+    T: TransmuteInto<NonEmptySlice<'static, u8>>,
+{
+    src.transmute_into()
+}
+
+evil_constructor(0u128).first() // muahaha!
+```
+This safety hazard is not materially different from the one that would be induced if the type author implemented `DerefMut<Target=NonEmptySliceInner>` for `NonEmptySlice`, or made the `private` module `pub`, or otherwise explicitly provided outsiders with unrestricted mutable access to `data`.
+
+##### Recommendation
+We recommend that that implementers of this RFC initially simplify constructability by:
+ - adopting our simplified definition of constructability
+ - demoting `NeglectStability` to unsafe status (i.e., not implementing `SafeTransmuteOptions` for `NeglectStability`; *only* `UnsafeTransmuteOptions`)
+
+If and when the implementation of `TransmuteFrom` encodes our complete definition of constructability, `NeglectStability` shall become a safe transmute option.
+
+
+## Dissecting Stability
+
+In this section, we iteratively review the design of the stability declaration traits, starting with a simplest-possible design:
+```rust
+pub unsafe trait PromiseTransmutableFrom
+{
+    type Archetype;
+}
+
+pub unsafe trait PromiseTransmutableInto
+{
+    type Archetype;
+}
+```
```

Comment:
Clarify the purpose of *Dissecting Stability*—it's about turning user-facing complexity into complexity the type system handles.

````suggestion
In this section, we iteratively review the design of the stability declaration traits, starting with a simplest-possible design:
```rust
pub unsafe trait PromiseTransmutableFrom
{
    type Archetype;
}

pub unsafe trait PromiseTransmutableInto
{
    type Archetype;
}
```
This section gives concrete examples of how misusing these simple, unsafe definitions can lead to unsoundness, and then refines them until they are immune to misuse (and, consequently, *safe*).
````